### PR TITLE
Implement support for National cloud deployments (Issue #937)

### PR DESCRIPTION
**Pull request by abraunegg**
* Implement support for National cloud deployments as per https://docs.microsoft.com/en-us/graph/deployments

### CODE REVIEW

**Comment by abraunegg** [comment]
@norbusan 
Whilst the code changes do work - that is ability to set the Auth / Graph URL's to specific objects, unable to test this myself to make sure that there are no other issues. 

Documentation for this feature 'looks' right based on existing application configuration, but again, no way to test (except for using new application id which we know works)

**Comment by abraunegg** [comment]
@norbusan 
This should be good for another review now. It is quite a large feature implementation now, including a DB schema update, but those changes ONLY ever get used / called if a National Cloud Deployment is configured for use.

When in use, the output is like this:
```
./onedrive --confdir '~/.config/onedrive-business/' --synchronize --verbose
Using 'user' Config Dir: /home/alex/.config/onedrive-business/
Using 'system' Config Dir: 
Configuration file successfully loaded
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/alex/OneDriveBusiness
Application version: v2.4.2-29-g58423d8
Account Type: business
Default Drive ID: b!bO8V7s9SSk6r7mWHpIjURotN33W1W2tEv3OXV_oFIdQimEdOHR-1So7CqeT1MfHA
Default Root ID: 01WIXGO5V6Y2GOVW7725BZO354PWSELRRZ
Remaining Free Space: 1098476449374
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 01WIXGO5V6Y2GOVW7725BZO354PWSELRRZ
Adding OneDrive root details for processing
Adding 5 OneDrive items for processing from OneDrive root
Adding 2 OneDrive items for processing from /SMPP
Processing 8 OneDrive items to ensure consistent local state
Uploading differences of .
Processing root
The directory has not changed
Processing random_files
The directory has not changed
Processing random_images
The directory has not changed
Processing random_videos
The directory has not changed
Processing newfile.txt
The file has not changed
Processing SMPP
The directory has not changed
Processing SMPP-IF-SPEC_v3_3-24858.pdf
The file has not changed
Processing SMPP_v3_4_Issue1_2-24857.pdf
The file has not changed
Uploading new items of .
Applying changes of Path ID: 01WIXGO5V6Y2GOVW7725BZO354PWSELRRZ
Adding OneDrive root details for processing
Adding 5 OneDrive items for processing from OneDrive root
Adding 2 OneDrive items for processing from /SMPP
Processing 8 OneDrive items to ensure consistent local state
```
Deleting data on OneDrive:
```
./onedrive --confdir '~/.config/onedrive-business/' --synchronize --verbose
Using 'user' Config Dir: /home/alex/.config/onedrive-business/
Using 'system' Config Dir: 
Configuration file successfully loaded
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/alex/OneDriveBusiness
Application version: v2.4.2-29-g58423d8
Account Type: business
Default Drive ID: b!bO8V7s9SSk6r7mWHpIjURotN33W1W2tEv3OXV_oFIdQimEdOHR-1So7CqeT1MfHA
Default Root ID: 01WIXGO5V6Y2GOVW7725BZO354PWSELRRZ
Remaining Free Space: 1098476449374
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 01WIXGO5V6Y2GOVW7725BZO354PWSELRRZ
Adding OneDrive root details for processing
Adding 4 OneDrive items for processing from OneDrive root
Processing 5 OneDrive items to ensure consistent local state
Trying to delete item SMPP
Deleting item SMPP
Uploading differences of .
Processing root
The directory has not changed
Processing random_files
The directory has not changed
Processing random_images
The directory has not changed
Processing random_videos
The directory has not changed
Processing newfile.txt
The file has not changed
Uploading new items of .
Applying changes of Path ID: 01WIXGO5V6Y2GOVW7725BZO354PWSELRRZ
Adding OneDrive root details for processing
Adding 4 OneDrive items for processing from OneDrive root
Processing 5 OneDrive items to ensure consistent local state
```

File: src/itemdb.d [hunk 1]
```diff
@@ -117,6 +118,7 @@ final class ItemDatabase
 				remoteDriveId    TEXT,
 				remoteId         TEXT,
 				deltaLink        TEXT,
+				syncStatus       TEXT,
```
Comment (norbusan) [inline src/itemdb.d:121]: Is it easier to keep the status as text? I mean, that is not the really optimal storage type for a Boolean, and we only save Y or N, right?

Comment (abraunegg) [reply src/itemdb.d:121]: Is it easier? Probably the same as boolean storage wise - as a bool would be stored as 0x00 / 0xFF (or whatever sqlite uses) vs 0x59 / 0x4E .. so same amount of memory space IMHO .. happy to look at changing that to bool tho if you think thats what should be used.

**Comment by abraunegg** [comment]
> Hi Alex,
> since I cannot test, I only read through the massive changes ... not that I can say from reading that this is semantically correct.
> 
> The is one comment about the storage size/type of the syncStatus, but that is not that important.

You should be able to test this - if you 'modify' manually sync.d:
```
private bool nationalCloudDeployment = false;
```
to 
```
private bool nationalCloudDeployment = true;
```

This will 'bypass' the checks where the flag gets set, and uses all the new code, but without using the National Azure Deployment ... it is how I have been testing.

File: src/itemdb.d (hunk 1 above)
Comment (norbusan) [reply src/itemdb.d:121]: Not that I really care, I'm just used to optimize db tables due to lots of big data stuff I am doing ;-) I'm fine with it if you are fine with it ;-)

**Pull request merged**